From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Fri, 8 Aug 2025 12:59:39 +0200
Subject: xdp-app-info: Add separated permissions ID function

While for flatpaks the app-ID matches the value used for permissions
store, in the snap world a snap may contain multiple applications and
they are not fully confined in between each other (processes are, but
they share common data paths), reason why instead of using per-app
permissions we use a per-snap permission.

The snap ID used for permissions is not matching any desktop file or
desktop ID, so this breaks the assumptions of the rest of the stack.

As per this, split the application ID with the permissions ID, by adding
another function that we will re-implement in next commits to have snap
specific behavior.

See: #155 and #769

Origin: https://github.com/flatpak/xdg-desktop-portal/pull/1764
---
 src/xdp-app-info.c | 11 +++++++++++
 src/xdp-app-info.h |  2 ++
 tests/xdp_utils.py |  5 +++++
 3 files changed, 18 insertions(+)

diff --git a/src/xdp-app-info.c b/src/xdp-app-info.c
index 64624bc..f6109c0 100644
--- a/src/xdp-app-info.c
+++ b/src/xdp-app-info.c
@@ -462,6 +462,17 @@ xdp_app_info_get_id (XdpAppInfo *app_info)
   return priv->id;
 }
 
+const char *
+xdp_app_info_get_permissions_id (XdpAppInfo *app_info)
+{
+  XdpAppInfoPrivate *priv;
+
+  g_return_val_if_fail (app_info != NULL, NULL);
+
+  priv = xdp_app_info_get_instance_private (app_info);
+  return priv->id;
+}
+
 const char *
 xdp_app_info_get_instance (XdpAppInfo *app_info)
 {
diff --git a/src/xdp-app-info.h b/src/xdp-app-info.h
index 0fedf88..76c6b22 100644
--- a/src/xdp-app-info.h
+++ b/src/xdp-app-info.h
@@ -57,6 +57,8 @@ gboolean xdp_app_info_is_host (XdpAppInfo *app_info);
 
 const char * xdp_app_info_get_id (XdpAppInfo *app_info);
 
+const char * xdp_app_info_get_permissions_id (XdpAppInfo *app_info);
+
 const char * xdp_app_info_get_instance (XdpAppInfo *app_info);
 
 const char * xdp_app_info_get_engine (XdpAppInfo *app_info);
diff --git a/tests/xdp_utils.py b/tests/xdp_utils.py
index fc6b392..e183591 100644
--- a/tests/xdp_utils.py
+++ b/tests/xdp_utils.py
@@ -262,6 +262,7 @@ class AppInfo:
 
     kind: AppInfoKind
     app_id: str
+    permissions_id: str
     desktop_file: str
     env: dict[str, str] = field(default_factory=dict)
     files: dict[Path, bytes] = field(default_factory=dict)
@@ -302,6 +303,7 @@ class AppInfo:
         return cls(
             kind=kind,
             app_id=app_id,
+            permissions_id=app_id,
             desktop_file=desktop_file,
             env=env,
             files=files,
@@ -372,6 +374,7 @@ enumerable-devices={usb_queries}
         return cls(
             kind=kind,
             app_id=app_id,
+            permissions_id=app_id,
             desktop_file=desktop_file,
             env=env,
             files=files,
@@ -428,6 +431,7 @@ DesktopFile={desktop_file}
         return cls(
             kind=kind,
             app_id=app_id,
+            permissions_id=app_id,
             desktop_file=desktop_file,
             env=env,
             files=files,
@@ -490,6 +494,7 @@ Network=shared
         return cls(
             kind=kind,
             app_id=app_id,
+            permissions_id=app_id,
             desktop_file=desktop_file,
             env=env,
             files=files,
